### PR TITLE
Add `rootParentIds` option to fit the case when all items contain a valid parentId string

### PR DESCRIPTION
**Pull request by rrbe**
See #23

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -64,6 +64,7 @@ You can provide a second argument to arrayToTree with configuration options. Rig
 - `childrenField`: key which will contain all child nodes of the parent node. Default: `"children"`
 - `dataField`: key which will contain all properties/data of the original items. Set to null if you don't want a container. Default: `"data"`
 - `throwIfOrphans`: option to throw an error if the array of items contains one or more items that have no parents in the array. This option has a small runtime penalty, so it's disabled by default. When enabled, the function will throw an error containing the parentIds that were not found in the items array. When disabled, the function will just ignore orphans and not add them to the tree. Default: `false`
+- `rootParentIds`: key array used to specify an item whose parent id is an orphan, but you don't want it and it's sub items to be ignored. It's useful when your tree is a subset of full tree, which means there is no item whose parent id is one of `undefined` / `null` / `''`. The array you pass in will be merged with the default value. See[#23](https://github.com/philipstanislaus/performant-array-to-tree/issues/23). Default: `[null, undefined, '']`
```
Comment (philipstanislaus) [inline README.md:67]:
```suggestion
- `rootParentIds`: Array of parent ids that should be considered the top or root elements of the tree. This is useful when your tree is a subset of full tree, which means there is no item whose parent id is one of `undefined`, `null` or `''`. The array you pass in will be merged with the default value. For more details, see [#23](https://github.com/philipstanislaus/performant-array-to-tree/issues/23). Default: `[null, undefined, '']`
```

File: src/arrayToTree.spec.ts [hunk 2]
```diff
@@ -216,6 +216,34 @@ describe('arrayToTree', () => {
...
   })
 
+  it('should treat objects with non-zero length string as parentId as root objects', () => {
+    expect(arrayToTree(
+      [
+        { id: '4', parentId: 'orphan1', custom: 'abc' },
+        { id: '31', parentId: '4', custom: '12' },
+        { id: '1941', parentId: '418', custom: 'de' },
+        { id: '1', parentId: '418', custom: 'ZZZz' },
+        { id: '418', parentId: 'orphan2', custom: 'ü' },
```
[the remark this answers is not in the document]
Comment (rrbe) [reply src/arrayToTree.spec.ts:226]: Actually, the `parentId: '13'` below this line of code is the `orphan3`, I will just rename it.

**Comment by rrbe** [comment]
> Thanks, left a few comments.
> 
> I feel that we should not merge the supplied array with the default, but replace it. Otherwise there is no fine grained control when to throw. What do you think?

These modifications are completed, and I also added test case about the `rootParentId` options replacement, please check it and see what else needs to be modified

File: src/arrayToTree.spec.ts [hunk 3]
```diff
@@ -309,6 +309,48 @@ describe('arrayToTree', () => {
     )
   })
 
+  it('should throw if a node has parentId that both exists in another node and is in rootParentIds', () => {
```
Comment (philipstanislaus) [inline src/arrayToTree.spec.ts:312]: Not sure I understand this test case - the description here reads as if it should throw, but below it does not. I think throwing an error would be the correct behavior. Is that possible without changing the runtime performance?

File: src/arrayToTree.spec.ts [hunk 4]
```diff
@@ -259,6 +287,70 @@ describe('arrayToTree', () => {
...
+            data: { id: 'bbb', parentId: 'aaa' }, children: [
+              { data: { id: 'ccc', parentId: 'bbb' }, children: [] },
+            ],
+          },
+        ],
+      },
+    ])
+  })
+
+  it('should default rootParentIds be replaced by the value user passd in', () => {
```
Comment (philipstanislaus) [inline src/arrayToTree.spec.ts:340]:
```suggestion
  it('should replace default rootParentIds by the provided value', () => {
```

File: src/arrayToTree.spec.ts (hunk 3 above)
Comment (rrbe) [reply src/arrayToTree.spec.ts:312]: Sorry for reply late, this is an embarrassing mistake...

I've changed the behavior of the test case, the related code has also made some change, please check `arrayToTree.ts   line 55 - line 58`

"Is that possible without changing the runtime performance", I'm not sure if I understand your concern, at first I thought we need to add some additional variables and a new loop to check whether the rootParentIds and all node ids have intersection.

But after careful thinking, I think we only need to check whether the id exists in rootParentIds in each round of the loop.

This is based on the following conditions, that is, a valid array is either an empty array, or it must contain a node whose parentId is one of rootParentIds([undefined, null, '', ...customized]). So if a node has parentId that both exists in another node and is in rootParentIds, it must throw an error.

BTW, my english is not so good, so I'm not sure if the error message is clear and easy to read

**Review by philipstanislaus — APPROVED** [review]
@rrbe sorry for the very late feedback, I just remembered there were some issues/PR still open here. 

I have thought about the architecture a bit again and changed the data type for `rootParentIds` to an object for better performance.

This should be fine now – I will merge it shortly.

Many thanks for your contribution and sorry again for the delay!

**Pull request merged**